File: dotnet/src/OpenQA.Selenium.Winium/DesktopOptions.cs
namespace OpenQA.Selenium.Winium
{
    #region using

    using System.Collections.Generic;

    using OpenQA.Selenium;
    using OpenQA.Selenium.Remote;

    #endregion

    /// <summary>
    /// The keyboard simulator type.
    /// </summary>
    public enum KeyboardSimulatorType
    {
        /// <summary>
        /// Based on SendKeys Class.<see href="https://msdn.microsoft.com/en-us/library/system.windows.forms.sendkeys(v=vs.110).aspx">See more</see>.
        /// </summary>
        BasedOnWindowsFormsSendKeysClass, 

        /// <summary>
        /// Based on Windows Input Simulator.
        /// For additional methods should be cast to the KeyboardSimulatorExt. <see href="http://inputsimulator.codeplex.com/">See more</see>.
        /// </summary>
        BasedOnInputSimulatorLib
    }

    /// <summary>
    /// Class to manage options specific to <see cref="WiniumDriver"/> 
    /// wich uses <see href="https://github.com/2gis/Winium.Desktop">Winium.Desktop</see>.
    /// </summary>
    public class DesktopOptions : IWiniumOptions
    {
        #region Constants

        private const string ApplicationPathOption = "app";

        private const string ArgumentsOption = "args";

        private const string DebugConnectToRunningAppOption = "debugConnectToRunningApp";

        private const string KeyboardSimulatorOption = "keyboardSimulator";

        private const string LaunchDelayOption = "launchDelay";

        #endregion

        #region Fields

        private string applicationPath;

        private string arguments;

        private bool? debugConnectToRunningApp;

        private KeyboardSimulatorType? keyboardSimulator;

        private int? launchDelay;

        #endregion

        #region Public Properties

        /// <summary>
        /// Gets or sets the absolute local path to an .exe file to be started. 
        /// This capability is not required if debugConnectToRunningApp is specified.
        /// </summary>
        public string ApplicationPath
        {
            set
            {
                this.applicationPath = value;
            }
        }

        /// <summary>
        /// Gets or sets startup argunments of the application under test.
        /// </summary>
        public string Arguments
        {
            set
            {
                this.arguments = value;
            }
        }

        /// <summary>
        /// Gets or sets a value indicating whether debug connect to running app.
        /// If true, then application starting step are skipped.
        /// </summary>
        public bool DebugConnectToRunningApp
        {
            set
            {
                this.debugConnectToRunningApp = value;
            }
        }

        /// <summary>
        /// Gets or sets the keyboard simulator type.
        /// </summary>
        public KeyboardSimulatorType KeyboardSimulator
        {
            set
            {
                this.keyboardSimulator = value;
            }
        }

        /// <summary>
        /// Gets or sets the launch delay in milliseconds, to be waited to let visuals to initialize after application started.
        /// </summary>
        public int LaunchDelay
        {
            set
            {
                this.launchDelay = value;
            }
        }

        #endregion

        #region Public Methods and Operators

        /// <summary>
        /// Convert options to DesiredCapabilities for Winium Desktop Driver 
        /// </summary>
        /// <returns>The DesiredCapabilities for Winium Desktop Driver with these options.</returns>
        public ICapabilities ToCapabilities()
        {
            var capabilityDictionary = new Dictionary<string, object>
                                           {
                                               { ApplicationPathOption, this.applicationPath }
                                           };

            if (!string.IsNullOrEmpty(this.arguments))
            {
                capabilityDictionary.Add(ArgumentsOption, this.arguments);
            }

            if (this.debugConnectToRunningApp.HasValue)
            {
                capabilityDictionary.Add(DebugConnectToRunningAppOption, this.debugConnectToRunningApp);
            }

            if (this.keyboardSimulator.HasValue)
            {
                capabilityDictionary.Add(KeyboardSimulatorOption, this.keyboardSimulator);
            }

            if (this.launchDelay.HasValue)
            {
                capabilityDictionary.Add(LaunchDelayOption, this.launchDelay);
            }

            return new DesiredCapabilities(capabilityDictionary);
        }

        #endregion
    }
}

File: java/src/main/java/org/openqa/selenium/winium/DesktopOptions.java
package org.openqa.selenium.winium;

import org.openqa.selenium.Capabilities;
import org.openqa.selenium.remote.DesiredCapabilities;

import java.util.HashMap;

/**
 * Class to manage options specific to {@link WiniumDriver}
 * which uses <a href="https://github.com/2gis/Winium.Desktop">Winium.Desktop</a>
 */
public class DesktopOptions implements WiniumOptions {
    private static final String APPLICATION_PATH_OPTION = "app";
    private static final String ARGUMENTS_OPTION = "args";
    private static final String DEBUG_CONNECT_TO_RUNNING_APP_OPTION = "debugConnectToRunningApp";
    private static final String KEYBOARD_SIMULATOR_OPTION = "keyboardSimulator";
    private static final String LAUNCH_DELAY_OPTION = "launchDelay";

    private String applicationPath;
    private String arguments;
    private Boolean debugConnectToRunningApp;
    private KeyboardSimulatorType keyboardSimulator;
    private Integer launchDelay;

    /**
     * Sets the absolute local path to an .exe file to be started.
     * This capability is not required if debugConnectToRunningApp is specified.
     * @param applicationPath Absolute local path to an .exe file to be started.
     */
    public void setApplicationPath(String applicationPath) {
        this.applicationPath = applicationPath;
    }

    /**
     * Sets startup argunments of the application under test.
     * @param arguments Startup argunments of the application under test.
     */
    public void setArguments(String arguments) {
        this.arguments = arguments;
    }

    /**
     * Sets a value indicating whether debug connect to running app.
     * If true, then application starting step are skipped.
     * @param debugConnectToRunningApp Value indicating whether debug connect to running app.
     */
    public void setDebugConnectToRunningApp(Boolean debugConnectToRunningApp) {
        this.debugConnectToRunningApp = debugConnectToRunningApp;
    }

    /**
     * Sets the keyboard simulator type.
     * @param keyboardSimulator Keyboard simulator type
     */
    public void setKeyboardSimulator(KeyboardSimulatorType keyboardSimulator) {
        this.keyboardSimulator = keyboardSimulator;
    }

    /**
     * Sets the launch delay in milliseconds, to be waited to let visuals to initialize after application started.
     * @param launchDelay Launch delay in milliseconds
     */
    public void setLaunchDelay(Integer launchDelay) {
        this.launchDelay = launchDelay;
    }

    /**
     * Convert options to DesiredCapabilities for Winium Desktop Driver
     * @return The DesiredCapabilities for Winium Desktop Driver with these options.
     */
    public Capabilities toCapabilities() {
        HashMap<String, Object> capabilityDictionary = new HashMap<String, Object>();
        capabilityDictionary.put(APPLICATION_PATH_OPTION, applicationPath);

        if ((arguments != null) && (arguments.length() > 0)) {
            capabilityDictionary.put(ARGUMENTS_OPTION, arguments);
        }

        if (debugConnectToRunningApp != null) {
            capabilityDictionary.put(DEBUG_CONNECT_TO_RUNNING_APP_OPTION, debugConnectToRunningApp);
        }

        if (keyboardSimulator != null) {
            capabilityDictionary.put(KEYBOARD_SIMULATOR_OPTION, keyboardSimulator);
        }

        if (launchDelay != null) {
            capabilityDictionary.put(LAUNCH_DELAY_OPTION, launchDelay);
        }

        return new DesiredCapabilities(capabilityDictionary);
    }
}
